staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/validator.go

```diff
@@ -0,0 +1,284 @@
+func (c *CELValidatorCompiler) DefinitionMatches(definition *v1alpha1.ValidatingAdmissionPolicy, a admission.Attributes, o admission.ObjectInterfaces) (bool, error) {
+	criteria := matchCriteria{constraints: definition.Spec.MatchConstraints}
+	return c.Matcher.Matches(&criteria, a, o)
+}
+
+// BindingMatches returns whether this ValidatingAdmissionPolicyBinding matches the provided admission resource request
+func (c *CELValidatorCompiler) BindingMatches(binding *v1alpha1.ValidatingAdmissionPolicyBinding, a admission.Attributes, o admission.ObjectInterfaces) (bool, error) {
+	if binding.Spec.MatchResources == nil {
+		return true, nil
+	}
+	criteria := matchCriteria{constraints: binding.Spec.MatchResources}
+	return c.Matcher.Matches(&criteria, a, o)
+}
+
+// ValidateInitialization checks if Matcher is initialized.
+func (c *CELValidatorCompiler) ValidateInitialization() error {
+	return c.Matcher.ValidateInitialization()
+}
+
+type validationActivation struct {
+	object    *unstructured.Unstructured
+	oldObject *unstructured.Unstructured
+	params    *unstructured.Unstructured
+	request   *unstructured.Unstructured
+}
+
+// ResolveName returns a value from the activation by qualified name, or false if the name
+// could not be found.
+func (a *validationActivation) ResolveName(name string) (interface{}, bool) {
+	switch name {
+	case ObjectVarName:
+		return a.object.Object, true
+	case OldObjectVarName:
+		return a.oldObject.Object, true
+	case ParamsVarName:
+		return a.params.Object, true
+	case RequestVarName:
+		return a.request.Object, true
+	default:
+		return nil, false
+	}
+}
+
+// Parent returns the parent of the current activation, may be nil.
+// If non-nil, the parent will be searched during resolve calls.
+func (a *validationActivation) Parent() interpreter.Activation {
+	return nil
+}
+
+// Compile compiles the cel expression defined in ValidatingAdmissionPolicy
+func (c *CELValidatorCompiler) Compile(p *v1alpha1.ValidatingAdmissionPolicy) (Validator, error) {
+	if len(p.Spec.Validations) == 0 {
+		return nil, nil
+	}
+	hasParam := false
+	if p.Spec.ParamKind != nil {
+		hasParam = true
+	}
+	compilationResults := make([]CompilationResult, len(p.Spec.Validations))
+	for i, validation := range p.Spec.Validations {
+		compilationResults[i] = CompileValidatingPolicyExpression(validation.Expression, hasParam)
+	}
+	return &CELValidator{policy: p, compilationResults: compilationResults}, nil
+}
+
+// CELValidator implements the Validator interface
+type CELValidator struct {
+	policy             *v1alpha1.ValidatingAdmissionPolicy
+	compilationResults []CompilationResult
+}
+
+func convertObjectToUnstructured(obj interface{}) *unstructured.Unstructured {
+	var unstructuredObj unstructured.Unstructured
+	jsonObj, _ := json.Marshal(obj)
+	_ = json.Unmarshal(jsonObj, &unstructuredObj.Object)
+
+	return &unstructuredObj
+}
+
+// Validate validates all cel expressions in Validator and returns PolicyDecisions and error.
```

Comment:
"returns a PolicyDecision for each CEL expression or returns an error"? 

Also, generally document the return structure better? I didn't realize until discussing this feature that a PolicyDecision can contain an ignored error using Kind=Allow and a non-nil Message. Also, should document under what circumstances an error is returned instead.